File: tree/104.c
/**
 * going recursive
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int maxd(struct TreeNode* root, int depth);

int maxDepth(struct TreeNode* root) {
    return maxd(root, 0);
}

int maxd(struct TreeNode* root, int depth)
{
    int depthleft, depthright;
    if(root==NULL) return depth;
    depthleft = maxd(root->left, depth+1);
    depthright = maxd(root->right, depth+1);
    return depthleft > depthright ? depthleft : depthright;
    
}

File: tree/104.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    var ret int
    MaxD(&ret, root, 0)
    return ret
}

func MaxD(ret *int, root *TreeNode, cur int) {
    if root.Left == nil && root.Right == nil {
        if cur + 1 > *ret {
            *ret = cur + 1
        }
    }
    if root.Left != nil {
        MaxD(ret, root.Left, cur+1)
    }
    if root.Right != nil {
        MaxD(ret, root.Right, cur+1)
    }
}
